mesonbuild/interpreter/interpreter.py

```diff
@@ -2701,6 +2707,26 @@ def check_clang_asan_lundef(self) -> None:
     # declare_dependency).
     def validate_within_subproject(self, subdir, fname):
         srcdir = Path(self.environment.source_dir)
+        builddir = Path(self.environment.build_dir)
+        if isinstance(fname, P_OBJ.DependencyVariableString):
+            def validate_installable_file(fname: str) -> bool:
+                installablefiles: T.Set[mesonlib.File] = set()
+                for d in self.build.data:
+                    for s in d.sources:
+                        installablefiles.add(s.absolute_path(srcdir, builddir))
+                installabledirs = [str(Path(srcdir, s.source_subdir)) for s in self.build.install_dirs]
+                if fname in installablefiles:
+                    return True
+                for d in installabledirs:
+                    if fname.startswith(d):
+                        return True
+                return False
```

Comment:
While this is technically correct, I'm a little bit worried that it's a lot of loops every time we create a File object. Can't tell if that's going to be a significant performance hit without profiling real world projects though.

Reply:
ah no sorry, only when picking files from subprojects... ok guess that's fine actually. Nevermind.

Reply:
Right, so I hope the impact is at least minimal. When you do pick files from a subproject, we loop over every installable file "to date" but I'm not sure how else to do it.

Ultimately I really want this done on time for the merge window, so if a clever performance boost is thought up after the fact we can merge that as a bugfix.